### PR TITLE
workspace: bump llvm / xla / stablehlo / pjrt plugins and restore tests xla@4663f04

### PR DESCRIPTION
**Pull request by hugomano**
- workspace: new module versions added to Bazel registry:
  - deps/xla: https://github.com/openxla/xla/commit/4663f04d2653d686833f9c306bd0d899c3127358 (`20241025.0-4663f04`) 
This version introduces an upgraded overlay with Python repository initialization for the TSL extension. The changes were necessitated by this [XLA commit](https://github.com/openxla/xla/commit/19b4ecbe98e2c2910d6ff6532cdb64636a71e6d5#diff-df5471accc0ccfdac45fb8397fb8e1488565baf112c72487912deda919680f8aR841), which triggers a repository call. Python's hermetic version is now provided via the .bazelrc file.
  - deps/stablehlo: https://github.com/openxla/stablehlo/commit/f7f8e4e35296deeff2e12e39421ac8d9599ba340 (`20241021.0-1c0b606`)  
  - deps/llvm: https://github.com/llvm/llvm-project/commit/f7f8e4e35296deeff2e12e39421ac8d9599ba340 (`20241022.0-6c4267f`)
 This version includes an upgraded module declaration with rules_python dependency added, as LLVM prepares its overlay for Bazel v8. [Relevant LLVM changes here](https://github.com/llvm/llvm-project/commit/3903cb4695012fb85a76c83b5616f2ffe6fa10f4)
- workspace: align used versions of rules_zig in the Bazel dependency graph
- deps/pjrt-artifacts: new PJRT artifacts [`v3.0.0`](https://github.com/zml/pjrt-artifacts/releases/tag/v3.0.0): 1ec65b4a55738f2d6f6de11bbdc2c86b1e8d03f1 uses [our our fork of XLA](https://github.com/zml/xla/commits/hugomano/xla-4663f04)
- pjrt: expose getPluginAttributes and getPluginAttribute private APIs and stableHLOCurrentVersion public API
- pjrt: updated customCallRegistry function with the new FFI provided by the PJRT API GPU extension 
Minimal support for this feature was added, as this is not the main focus of the current PR. ([See related changes in XLA here](https://github.com/openxla/xla/commit/ef85a7bcc308313492ebc50295a8a08b4e51b8f5))
- dialects/mlir/stablehlo: expose getCurrentVersion API and fix broken stablehloVersionFromCompatibilityRequirement API
- zml/pjrtx: serialize IR to the PJRT plugin's supported HLO version. If the plugin attribute is unavailable, it falls back to the compatibility requirement StableHLO version.
- zml/tensor: tests using batching dims for gather and scatter are now re-enabled
- zml/testing: use autoPlatform to run tests instead of forcing to CPU, allowing us to tests on differents platforms

### CODE REVIEW

File: mlir/dialects/stablehlo.zig [hunk 1]
```diff
@@ -1239,19 +1239,57 @@ pub const RngAlgorithm = struct {
...
+            str = buf[0..stream.pos];
+        }
     };
-    var context = Context{};
 
-    c.stablehloVersionFromCompatibilityRequirement(requirement, (struct {
-        pub fn callback(mlir_str: c.MlirStringRef, userdata: ?*anyopaque) callconv(.C) void {
-            const inner_ctx: *Context = @ptrCast(@alignCast(userdata));
-            inner_ctx.str = mlir.fromStringRef(mlir_str);
+    state.req = requirement;
```
[the remark this answers is not in the document]
Comment (gwenzek) [reply mlir/dialects/stablehlo.zig:1264]: what's "requirement" what is this code supposed to do if called with different requirements ?

[the remark this answers is not in the document]
Comment (hugomano) [reply mlir/dialects/stablehlo.zig]: I removed the `once` stuff

File: mlir/dialects/stablehlo.zig [hunk 2]
```diff
@@ -1239,19 +1239,57 @@ pub const RngAlgorithm = struct {
...
+            var stream = std.io.fixedBufferStream(&buf);
+            var context = .{ .writer = stream.writer() };
+            const WriterContext = @TypeOf(context);
+
+            c.stablehloGetCurrentVersion((struct {
+                pub fn callback(mlir_str: c.MlirStringRef, userdata: ?*anyopaque) callconv(.C) void {
+                    const inner_ctx: *WriterContext = @ptrCast(@alignCast(userdata));
+                    _ = inner_ctx.writer.write(mlir.fromStringRef(mlir_str)) catch unreachable;
+                }
+            }).callback, &context);
```
[the remark this answers is not in the document]
Comment (hugomano) [reply mlir/dialects/stablehlo.zig]: Commited suggested changes

File: examples/llama/main.zig [hunk 3]
```diff
@@ -119,7 +119,7 @@ pub fn asyncMain() !void {
         layer_start: u8 = 0,
         num_layers: ?u8 = null,
         seq_len: u32 = 256,
-        topk: u32 = 2,
+        topk: u32 = 1,
```
Comment (gwenzek) [inline examples/llama/main.zig:122]: not sure this should be commited ?

**Review by gwenzek — APPROVED** [review]

File: examples/llama/main.zig (hunk 3 above)
Comment (hugomano) [reply examples/llama/main.zig]: Thanks, I reverted

**Comment by gwenzek** [comment]
ouray !

**Pull request merged**